Fix #533 - Improve documentation and support for failover (#536)
irrdnet/irrd: docs/admins/availability-and-migration.rst

====================================
IRRd availability and data migration
====================================

This document explains approaches you can take to set up standby instances
of IRRd, how you might switch between instances, and share load between multiple
instances. This document only applies when you run an authoritative IRR
registry. If you do not, availability is much simpler: run multiple instances

> that each mirrors from the original sources.

Author: ```suggestion
that each mirror from the original sources.
```

You need to migrate or continuously mirror authoritative IRR data between your
IRRd instances in a number of cases:

* You are migrating to IRRDv4 from a legacy version of IRRd.
* You are migrating from one IRRDv4 deployment to the other.
* You are using one IRRd instance as the active instance, and would like to
  have a second on standby to promote to the active instance with the
  most recent data.

This document mainly discusses three kinds of IRRd instances:

* An **active** instance is an IRRd instance that is accepting and processing
  authoritative changes. Others can mirror from the active instance.
  This might also be a current legacy IRRd instance.
* A **standby** instance mirrors from the active instance, and is intended to
  be promoted to the active instance as part of a migration or fallback.
  While in a standby role, it can be used for queries, as those are read-only.
* A **query-only** instance mirrors from an active instance, and is never
  intended to be promoted to an active instance.

.. warning::
    It is important that there is only a single IRRd instance
    that processes authoritative changes, and is the single source of truth,

>     at the same time. IRRd does not support having multiple active instances.

Author: ```suggestion
    at one point in time. IRRd does not support having multiple active instances.
```

This document suggest three different approaches for configuring this,
each with their own upsides and downsides.

Option 1: using exports and NRTM for migrations and standby instances
---------------------------------------------------------------------
The first option is to use the same :doc:`mirroring </users/mirroring>`
features as any other kind of IRR data mirroring. This means using the files
placed in ``sources.{name}.export_destination`` by the active instance
as the ``sources.{name}.import_source`` for the standby instances,
and having standby's follow the active NRTM stream.
If you are migrating from a legacy version of IRRd, this is most likely your
only option.

This is a bit different from "regular" mirroring, where the mirror

> is never meant to be promoted to an active instance, and may be run by entirely

Author: ```suggestion
is never meant to be promoted to an active instance, and instances may be run by entirely
```

different organisations for different reasons.
There are a number of important special circumstances when using exports and
NRTM for migrations or availability, which are detailed below.

Note that an active IRRd instance for one IRR registry may simultaneously be a

> regular mirror for other sources.

Author: ```suggestion
regular mirror for other registries.
```

keys are already in the keychain, and safe to run while IRRd is running.

Password hashes
~~~~~~~~~~~~~~~
Password authentication depends on password hashes in `mntner` objects.
To improve security, these password hashes are not included in exports or
NRTM streams for regular mirrors in IRRDv4.

However, when an IRRd mirror is a standby
instance that may need to take an active role later, it needs all password
hashes. To support this, you need to configure a special mirroring process
on the current active instance:

* Set ``sources.{name}.export_destination_unfiltered`` to a path where IRRd
  will store exports that include full password hashes. Other than including
  full hashes, this works the same as ``sources.{name}.export_destination``.
  Then, distribute those files to your standby instance, and point
  ``import_source`` to their location.
* Set ``sources.{name}.nrtm_access_list_unfiltered`` to an access list defined
  in the configuration file. Any IP on this access list will receive
  full password hashes when doing NRTM requests. Other than that, NRTM works
  identical to filtered queries. Set this to the IPs of your standby instances.

> If you are migrating from a different IRR server, make sure that password

Author: ```suggestion
If you are migrating from a different IRR server, check that password
```

hashes are not filtered.

Serials
~~~~~~~
Each IRRd instance potentially creates its own set of NRTM serials when
importing changes over NRTM.
This means that when switching to a different instance, mirrors would
have to refresh their data.

IRRd can run a mirror in synchronised serial mode. This is used by some
deployments to spread their query load over multiple read-only instances.
For further details, see the
:ref:`NRTM serial handling documentation <mirroring-nrtm-serials>`.

> Do note that without synchronised serials, NRTM users must not be switched

Author: better wording
